module/zcommon/zfs_fletcher.c

```diff
@@ -245,11 +214,258 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+extern struct fletcher_4_calls fletcher_4_avx2_calls;
+
+static const struct fletcher_4_calls *fletcher_4_algos[] = {
+	&fletcher_4_generic_calls,
+#if defined(HAVE_AVX) && defined(HAVE_AVX2)
+	&fletcher_4_avx2_calls,
+#endif
+};
+
+#define	ZFS_ARRAY_SIZE(a) 	(sizeof (a) / sizeof (*a))
+
+static const struct fletcher_4_calls *
+    fletcher_4_chosen = &fletcher_4_generic_calls;
+static const struct fletcher_4_calls *fletcher_4_fastest = NULL;
+
+static enum fletcher_selector {
+	FLETCHER_FASTEST = 0,
+	FLETCHER_ORIGINAL,
+	FLETCHER_CYCLE
+} fletcher_4_impl_seletor = FLETCHER_ORIGINAL;
+
+static const char *selector_names[] = {
+	"fastest",
+	"original",
+#if !defined(_KERNEL)
+	"cycle",
+#endif
+};
+
+static kmutex_t fletcher_4_impl_seletor_lock;
+
+static kstat_t *fletcher_4_kstat;
+
+static kstat_named_t fletcher_4_kstat_data[ZFS_ARRAY_SIZE(fletcher_4_algos)];
+
+int
+fletcher_4_impl_set(const char *selector)
+{
+	unsigned idx, i;
+
+	for (i = 0; i < ZFS_ARRAY_SIZE(selector_names); i++) {
+		if (strncmp(selector, selector_names[i],
+		    strlen(selector_names[i])) == 0) {
+			idx = i;
+			break;
+		}
+	}
+	if (i >= ZFS_ARRAY_SIZE(selector_names))
+		return (-EINVAL);
+
+	if (fletcher_4_impl_seletor == idx)
+		return (0);
+
+	mutex_enter(&fletcher_4_impl_seletor_lock);
+	switch (idx) {
+	case FLETCHER_FASTEST:
+		fletcher_4_chosen = fletcher_4_fastest;
+		break;
+	case FLETCHER_ORIGINAL:
+	case FLETCHER_CYCLE:
+		fletcher_4_chosen = &fletcher_4_generic_calls;
+		break;
+	}
+
+	fletcher_4_impl_seletor = idx;
+	mutex_exit(&fletcher_4_impl_seletor_lock);
+
+	return (0);
+}
+
+static inline const struct fletcher_4_calls *fletcher_4_impl_get(void)
+{
+#if !defined(_KERNEL)
+	if (fletcher_4_impl_seletor == FLETCHER_CYCLE) {
+		static unsigned int cycle_count = 0;
+		const struct fletcher_4_calls *algo = NULL;
+		unsigned int index;
+
+		while (1) {
+			index = __sync_fetch_and_add(&cycle_count, 1);
+			algo = fletcher_4_algos[
+			    index % ZFS_ARRAY_SIZE(fletcher_4_algos)];
+			if (algo->valid == NULL || algo->valid())
+				break;
+		}
+		return (algo);
+	}
+#endif
+	asm volatile("":::"memory");
+	return (fletcher_4_chosen);
+}
+
+#define	BENCH_SIZE	4096
+
+void
+fletcher_4_init(void)
+{
+	static char databuf[BENCH_SIZE];
+	const uint64_t const bench_ns = (50 * 1000 * 1000ULL); /* 50ms */
+	kstat_named_t *stat;
+	unsigned long best_run_count = 0;
+	unsigned long best_run_index = 0;
+	int i;
+
+	stat = &fletcher_4_kstat_data[0];
+	for (i = 0; i < ZFS_ARRAY_SIZE(fletcher_4_algos); i++, stat++) {
```

Comment:
I'd prefer to avoid rely on incrementing a pointer like this in the for loop.  It's a little to clever for easy readability.  Let's just do the simple thing below.

```
stat = &fletcher_4_kstat_data[i];
```